### ISSUE TITLE
why hbase10 create so many thread

### ISSUE THREAD

**Issue by zengweizhan** [brianfrankcooper/YCSB#662]
### I used ycsb to test performance between  hbase094 and hbase1.1.2 , but when i used hbase10 , it would create too many thread that couse  OOM , I donit know why?
here is my test:

1. python ./bin/ycsb load hbase094 -P workloads/write_workload -p threadcount=50
top -b -n 1 -H | grep test | grep java | wc -l
185

2. python ./bin/ycsb load hbase10  -P workloads/write_workload -p threadcount=50
top -b -n 1 -H | grep test | grep java | wc -l
13079

**Comment by busbey** [comment]
What version of YCSB are you using?

**Comment by zengweizhan** [comment]
both 0.50 and 0.70 has this question. but i used 098 to test hbaes1.1.2 the threads is normal.

**Comment by busbey** [comment]
In version 0.7.0 we switched to using a single connection object, so I'd expect the thread count to drop.

Just to clarify, you're saying that when the 0.7.0 `hbase10` client is run against an HBase 0.98 cluster the high thread count still happens, but if you use the same to talk to an HBase 1.1.2 cluster then it doesn't?

**Comment by zengweizhan** [comment]
As we know, in HBase,  the version 0.98 client can talk to version 1.1.2 server. I do a test by useing hbase098  talk to HBase 1.1.2 cluster, the high thread count problem was not happen.

> python ./bin/ycsb load hbase098 -P workloads/write_workload -p threadcount=50
top -b -n 1 -H | grep test | grep java | wc -l
182

I  test this in 0.50

**Comment by busbey** [comment]
AFAIK, the high thread count for the `hbase10` binding was due to us originally doing 1 connection per client thread. That should have been fixed in 0.7.0.

Please make sure your test is using 0.7.0 to compare the `hbase098` and `hbase10` bindings.

**Comment by zengweizhan** [comment]
I test my case in 0.7.0 , the problem is still there. 

in  **hbase098** , the thread count is 339
in  **hbase10** , the thread count is 12356

if you have not met the problem,  Can you give me some idea to find the differences .

**Comment by busbey** [comment]
I suppose I'd start investigating by using jstack and hoping the thread purposes could be determined by names.

**Comment by manolama** [comment]
There is a bug in the Hbase10 thread init that allows it to create multiple connection objects. I  submitted a patch here: https://github.com/brianfrankcooper/YCSB/pull/651

**Comment by zengweizhan** [comment]
I used 1  thread to  test  hbase10  in 0.7.0  ,   In jstack i   see many connection thread 

 > cat thread.info | grep hconnection | wc -l   
256

**Comment by busbey** [comment]
how many region servers do you have?

**Comment by zengweizhan** [comment]
3 region servers in my  test cluster.

**Comment by busbey** [comment]
oof. that is a lot of threads. Have you taken a look at jstack to get an idea of where they're coming from?

**Comment by zengweizhan** [comment]
i create hbase table in one region , so actually it should have one connection .....
i look my jstack info , most thread  info  are like : 

> "hconnection-0x1a48563c-shared--pool1-t186" daemon prio=10 tid=0x00007f698c668800 nid=0x1016 waiting on condition [0x00007f69fd2d1000]
   java.lang.Thread.State: TIMED_WAITING (parking)
    at sun.misc.Unsafe.park(Native Method)
    - parking to wait for  <0x000000060aa1e460> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
    at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:226)
    at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2082)
    at java.util.concurrent.LinkedBlockingQueue.poll(LinkedBlockingQueue.java:467)
    at java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1068)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)

And just  one thread info seen  right connection:
>
"hconnection-0x1a48563c-shared--pool1-t185" daemon prio=10 tid=0x00007f698c667000 nid=0x1015 in Object.wait() [0x00007f69fd3d2000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
    at java.lang.Object.wait(Native Method)
    at org.apache.hadoop.hbase.ipc.RpcClientImpl.call(RpcClientImpl.java:1243)
    locked <0x000000075f556038> (a org.apache.hadoop.hbase.ipc.Call)
    at org.apache.hadoop.hbase.ipc.AbstractRpcClient.callBlockingMethod(AbstractRpcClient.java:216)
    at org.apache.hadoop.hbase.ipc.AbstractRpcClient$BlockingRpcChannelImplementation.callBlockingMethod(AbstractRpcClient.java:300)
    at org.apache.hadoop.hbase.protobuf.generated.ClientProtos$ClientService$BlockingStub.multi(ClientProtos.java:31937)
    at org.apache.hadoop.hbase.client.MultiServerCallable.call(MultiServerCallable.java:128)
    at org.apache.hadoop.hbase.client.MultiServerCallable.call(MultiServerCallable.java:53)
    at org.apache.hadoop.hbase.client.RpcRetryingCaller.callWithoutRetries(RpcRetryingCaller.java:200)
    at org.apache.hadoop.hbase.client.AsyncProcess$AsyncRequestFutureImpl$SingleServerRequestRunnable.run(AsyncProcess.java:701)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)

**Comment by zengweizhan** [comment]
why i run 30 thread to insert hbase1.1.2  is faster than 200 thread.

**Comment by kruthar** [comment]
Since this is a different issue and this one is already closed can you open a new issue?

**Comment by risdenk** [comment]
Is this issue related to PR #651? It was just merged into master and may fix the OOM error?
